File: cpp/src/Parrot.cpp
#include <stdexcept>
#include <algorithm>

#include "Parrot.h"

using namespace std;

Parrot::Parrot(ParrotType parrotType, int numberOfCoconuts, double voltage, bool isNailed)
    : parrotType(parrotType), numberOfCoconuts(numberOfCoconuts), voltage(voltage), isNailed(isNailed) {
}

double Parrot::getSpeed() {
    switch (parrotType) {
        case EUROPEAN:
            return getBaseSpeed();
        case AFRICAN:
            return max(0.0, getBaseSpeed() - getLoadFactor() * numberOfCoconuts);
        case NORWEGIAN_BLUE:
            return (isNailed) ? 0 : getBaseSpeed(voltage);
        default:
            throw std::invalid_argument("Should be unreachable");
    }
}

const string Parrot::getCry() {
    std::string cry;
    switch (parrotType) {
        case EUROPEAN:
            return "Sqoork!";
        case AFRICAN:
            return "Squaark!";
        case NORWEGIAN_BLUE:
            return (voltage > 0) ? "Bzzzzzz" : "...";
        default:
            throw std::invalid_argument("Should be unreachable");
    }
}

double Parrot::getBaseSpeed() {
    return 12.0;
}

double Parrot::getLoadFactor() {
    return 9.0;
}

double Parrot::getBaseSpeed(double current_voltage) {
    return min(24.0, current_voltage * getBaseSpeed());
}

File: JavaScript/src/parrot.js
export const PARROT_TYPES = {
    EUROPEAN:       'EUROPEAN',
    AFRICAN:        'AFRICAN',
    NORWEGIAN_BLUE: 'NORWEGIAN_BLUE',
};

export class Parrot {
    constructor(type, numberOfCoconuts, voltage, isNailed) {
        this.type = type;
        this.numberOfCoconuts = numberOfCoconuts;
        this.voltage = voltage;
        this.isNailed = isNailed;
    }

    getSpeed() {
        switch (this.type) {
            case PARROT_TYPES.EUROPEAN:
                return this.getBaseSpeed();
            case PARROT_TYPES.AFRICAN:
                return Math.max(0, this.getBaseSpeed() - this.getLoadFactor() * this.numberOfCoconuts);
            case PARROT_TYPES.NORWEGIAN_BLUE:
                return (this.isNailed) ? 0 : this.getBaseSpeedWithVoltage(this.voltage);
        }
        throw new Error("Should be unreachable");
    }

    getBaseSpeedWithVoltage(voltage) {
        return Math.min(24, voltage * this.getBaseSpeed());
    }

    getLoadFactor() {
        return 9;
    }

    getBaseSpeed() {
        return 12;
    }

    getCry() {
        switch (this.type) {
            case PARROT_TYPES.EUROPEAN:
                return "Sqoork!";
            case PARROT_TYPES.AFRICAN:
                return "Sqaark!";
            case PARROT_TYPES.NORWEGIAN_BLUE:
                return this.voltage > 0 ? "Bzzzzzz" : "...";
        }
        throw new Error("Should be unreachable");
    }
}
